src/index.d.ts

```diff
@@ -132,67 +141,81 @@ export type TVScreenPropsValue<
       } & ClassProp<ClassValue>
   : V extends undefined
   ? {
       [K in keyof EV]?: isTrueOrArray<C["responsiveVariants"]> extends true
-        ? StringToBoolean<keyof EV[K]> | TVScreenPropsValue<EV, K, C>
+        ? StringToBoolean<keyof EV[K]> | TVScreenPropsValue<EV, ES, K, C>
         : StringToBoolean<keyof EV[K]>;
     } & ClassProp<ClassValue>
   : {
       [K in keyof V | keyof EV]?: isTrueOrArray<C["responsiveVariants"]> extends true
-        ? StringToBoolean<keyof V[K] | keyof EV[K]> | TVScreenPropsValue<EV & V, K, C>
-        : StringToBoolean<keyof V[K] | keyof EV[K]>;
+        ?
+            | (K extends keyof V ? StringToBoolean<keyof V[K]> : never)
+            | (K extends keyof EV ? StringToBoolean<keyof EV[K]> : never)
+            | TVScreenPropsValue<EV & V, S, K, C>
+        :
+            | (K extends keyof V ? StringToBoolean<keyof V[K]> : never)
+            | (K extends keyof EV ? StringToBoolean<keyof EV[K]> : never);
     } & ClassProp<ClassValue>;
 
 export type TVVariantKeys<V extends TVVariants<S>, S extends TVSlots> = V extends Object
   ? Array<keyof V>
   : undefined;
 
-export type TVReturnProps<V extends TVVariants<S>, S extends TVSlots, B extends ClassValue> = {
+export type TVReturnProps<
+  V extends TVVariants<S>,
+  S extends TVSlots,
+  B extends ClassValue,
+  EV extends TVVariants<ES>,
+  ES extends TVSlots,
+> = {
   base: B;
   slots: S;
   variants: V;
-  defaultVariants: TVDefaultVariants<V, S>;
-  compoundVariants: TVCompoundVariants<V, S, B>;
+  defaultVariants: TVDefaultVariants<V, S, EV, ES>;
+  compoundVariants: TVCompoundVariants<V, S, B, EV, ES>;
   compoundSlots: TVCompoundSlots<V, S, B>;
   variantKeys: TVVariantKeys<V, S>;
 };
 
 export type TVReturnType<
   V extends TVVariants<S>,
-  EV extends TVVariants,
   S extends TVSlots,
-  ES extends TVSlots,
   B extends ClassValue,
   C extends TVConfig<V, EV>,
+  EV extends TVVariants<ES>,
+  ES extends TVSlots,
 > = {
-  (props?: TVProps<V, EV, S, C>): ES extends undefined
+  (props?: TVProps<V, S, C, EV, ES>): ES extends undefined
     ? S extends undefined
       ? string
       : {[K in TVSlotsWithBase<S, B>]: (slotProps?: ClassProp) => string}
     : {[K in TVSlotsWithBase<ES & S, B>]: (slotProps?: ClassProp) => string};
-} & TVReturnProps<V, S, B>;
+} & TVReturnProps<V, S, B, EV, ES>;
 
 export type TV = {
   <
-    V extends TVVariants<S, B, EV> = undefined,
-    CV extends TVCompoundVariants<V, EV, S, B> = undefined,
-    DV extends TVDefaultVariants<V, EV, S> = undefined,
-    C extends TVConfig<V, EV> = undefined,
+    V extends TVVariants<S, B, EV>,
+    CV extends TVCompoundVariants<V, S, B, EV, ES>,
+    DV extends TVDefaultVariants<V, S, EV, ES>,
+    C extends TVConfig<V, EV>,
     B extends ClassValue = undefined,
     S extends TVSlots = undefined,
-    E extends ReturnType<TV> = undefined,
+    // @ts-expect-error
+    E extends TVReturnType = undefined,
+    // @ts-expect-error
```

Comment:
These two types basically are a recursive generics, and while I tried a few things, all the attempts I made simply resulted in breaking the types in real world usage (`tv.test.ts`).